File: leetcode/easy/141/141.rb
# Runtime: 56 ms, faster than 92.38% 
# Memory Usage: 210.5 MB, less than 67.62%
#
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @return {Boolean}
def hasCycle(head)
  return false if !head or !head.next

  tortoise = head.next
  hare = head.next.next

  0.upto(10000) do |_|
    return true if tortoise == hare

    tortoise = tortoise.next
    return false if !tortoise

    hare = hare.next
    return false if !hare or !hare.next
    hare = hare.next
  end

  false
end

File: leetcode/easy/141/141.js
/** 
 * Runtime: 88 ms, faster than 62.31% 
 * Memory Usage: 41.4 MB, less than 22.21%
 */ 

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
var hasCycle = function(head) {
  if (!head || !head.next) return false;

  let tortoise = head.next;
  let hare = head.next.next;

  for(let i=0; i<10001; ++i) {
    if (tortoise === hare) return true;
    tortoise = tortoise.next;
    if(!tortoise) return false;
    hare = hare.next;
    if(!hare || !hare.next) return false;
    hare = hare.next;
  }

  return false;
};
